node/core/approval-voting/src/approval_db/v2/migration_helpers.rs

```diff
@@ -0,0 +1,268 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Approval DB migration helpers.
+use super::{StoredBlockRange, *};
+use crate::backend::Backend;
+use polkadot_node_primitives::approval::v1::{
+	AssignmentCert, AssignmentCertKind, VrfOutput, VrfProof, VrfSignature, RELAY_VRF_MODULO_CONTEXT,
+};
+use polkadot_node_subsystem_util::database::Database;
+use std::{collections::HashSet, sync::Arc};
+
+use ::test_helpers::dummy_candidate_receipt;
+
+fn dummy_assignment_cert(kind: AssignmentCertKind) -> AssignmentCert {
+	let ctx = schnorrkel::signing_context(RELAY_VRF_MODULO_CONTEXT);
+	let msg = b"test-garbage";
+	let mut prng = rand_core::OsRng;
+	let keypair = schnorrkel::Keypair::generate_with(&mut prng);
+	let (inout, proof, _) = keypair.vrf_sign(ctx.bytes(msg));
+	let out = inout.to_output();
+
+	AssignmentCert { kind, vrf: VrfSignature { output: VrfOutput(out), proof: VrfProof(proof) } }
+}
+fn make_block_entry(
+	block_hash: Hash,
+	parent_hash: Hash,
+	block_number: BlockNumber,
+	candidates: Vec<(CoreIndex, CandidateHash)>,
+) -> BlockEntry {
+	BlockEntry {
+		block_hash,
+		parent_hash,
+		block_number,
+		session: 1,
+		slot: Slot::from(1),
+		relay_vrf_story: [0u8; 32],
+		approved_bitfield: make_bitvec(candidates.len()),
+		candidates,
+		children: Vec::new(),
+	}
+}
+
+fn make_bitvec(len: usize) -> BitVec<u8, BitOrderLsb0> {
+	bitvec::bitvec![u8, BitOrderLsb0; 0; len]
+}
+
+pub fn dummy_assignment_bitfield() -> CoreBitfield {
+	vec![
+		CoreIndex(0),
+		CoreIndex(1),
+		CoreIndex(2),
+		CoreIndex(3),
+		CoreIndex(4),
+		CoreIndex(5),
+		CoreIndex(6),
+		CoreIndex(7),
+	]
+	.try_into()
+	.expect("If failed, `CoreBitfield` is broken; qed")
+}
+
+/// Migrates `OurAssignment`, `CandidateEntry` and `ApprovalEntry` to version 2.
+/// Returns on any error.
+/// Must only be used in parachains DB migration code - `polkadot-service` crate.
+pub fn migrate_approval_db_v1_to_v2(db: Arc<dyn Database>, config: Config) -> Result<()> {
+	let mut backend = crate::DbBackend::new(db, config);
+	let all_blocks = backend
+		.load_all_blocks()
+		.map_err(|e| Error::InternalError(e))?
+		.iter()
+		.filter_map(|block_hash| {
+			backend.load_block_entry(block_hash).map_err(|e| Error::InternalError(e)).ok()?
```

Comment:
This would scale with the amount of unfinalized blocks (and forks). If finality lag goes past 50 blocks (IIRC) we backoff from authoring new blocks, so does the potential for this to get too big.

Since we are running this at startup, I don't think we can have so many blocks here that it would OOM. This is a migration, so after it's done, it won't run the second time. 

If you have a hard preference, I can break this up into a loop.

Reply:
Worth noting that this seems to discard the error and continue - is that intentional?

`.ok().ok().flatten()` does the same thing as far as I see.